Reviews component

Nick-Zabalza opened the pull request:
@GrumpyCat0709  @margotTenenbaum 

Hi Kate and Elizabeth,

It is finally here.

Sorry for the size but I kept running into unexpected hurdles. Like:

1. The drop down menus not being actual HTML drop down menus. Instead, they are custom scripts and html elements. This caused many issues but the one that caused the most bloat was that I also had to program the behavior of each menu. And, since I'm using redux, this meant putting all state variables in the store. Three state variables per drop down menu, just to control behavior. Each then needs its own reducer and its own action creator and at least one event handler. Then everything needs to be tested. I hate my drop down menus. But they work perfectly, damn it.

2. Speaking of tests. Another reason for the bloat is having to switch over to Enzyme/Jest. This involved rewriting some of the old tests. But it also caused significant bloat since, for some stupid reason I'm sure, Enzyme requires a lot of wrapper.update() and wrapper.render() statements. How did this setup ever become an industry standard?

3. I also realized I missed an important field in my database data. Which then involved updating the seed files and tests, too. Speaking of which, MAKE SURE YOU drop PTCReviewsService from your MongoDB instance and reseed your database or else you will have tons of errors.

4. Just in general, a lot of hidden features and interactivity for this component. Particularly the sorting algorithm for 'Most Helpful' reviews. For some strange reason, PetCo desicded to make this algorithm super complex. If you have the time, you can check out the comments - and the algorithm - in client/src/Components/Reviews/calculateFilteredReviews.jsx

One last note. Currently, the ReadMe talks about how you need to make alterations to run this service in developer mode. (Bullet point 4 of the "From project's root folder" section). As for now, you don't have to worry about this. It is already in developer mode. Instead, i have to worry about converting the code back to service mode before merging this PR.

margotTenenbaum commented:
Hi @Nick-Zabalza,

A truly impressive amount of work here, and a great job overall. Your README instructions were clear, and all tests passed. I clicked around the new component pretty thoroughly and compared it against PetCo's page, and you've captured all of the details really well. There was one bug that I found, which I will elaborate on below, but otherwise everything looks really good.

-There is a really strange bug where when I click on a yes or no on a review (turning the colors green/red) a random different review will also have its colors turned green/red when you click the arrow to move to the next page of reviews. I *think* it might be the review that ends up in a similar spot on the screen, but that wasn't always consistent.
I found the related tests (which passed) and they both seem to be happy path tests, and since the colors do change when a yes or no is clicked, they weren't able to capture this other strange behavior. Once you discover the cause of this bug, I might recommend adding a test to check that if a yes/no button is clicked, no others change colors.

-Another note on tests: some of your file names made it tricky to locate the yes/no button tests i was looking for. Specifically having reviewsIndex parts 1-4. I might recommend renaming these to make it easier for someone to know which file covers which general sections of the component (ie something like reviewHeader or reviewButtons). Alternately, maybe a comment at the top of each file to list the kinds of tests found therein.

-typo on line 39 of service.jsx (standalone)

Nick-Zabalza commented:
@margotTenenbaum 
Good catch!

Weird bug but, after I tracked down the source, not too surprising. The actual components were hardcoded to avoid this by checking if the specific review objects should have colors or not. However, when either yes or no button is clicked, it updates the review object directly. Which I thought would be fine since it is just a reference to the object being stored in the array.

However, in React, you have to use setState() to properly update a component's state. Except...none of my components have state I can directly update with setState since I'm using Redux...which then means the proper way to update that object is to have a reducer make the change(since Redux then handles calling setState appropriately). Which is now true. All I did was make it so the appropriate reducer updates the appropriate review object. No more issues.

I also updated tests to reflect this change and added a new test to test for this bug. All tests pass.

As for how the tests are ordered, I take no responsibility. I blame enzyme/jest. When I first started writing tests for this PR, I had a vision to keep things nice and orderly. A vision I would have easily been able to carry out had I been still writing the tests in the browser. But, at about every twist and turn, either jest, enzyme, or both didn't me to do what I expected to be able to do and so tests just started to lump together in no particular order; just to satisfy some weird quirk of the setup. That is also why there might not be as many non-happy path tests...enzyme/jest itself is the antipattern and I wrote the bare minimum number of tests so I could stop using enzyme/jest ASAP.

Thanks again for the stellar review, Elizabeth.
